pkg/controller/pxc/upgrade.go

```diff
@@ -115,7 +110,157 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
 	currentSet.Spec.Template.Spec.Affinity = pxc.PodAffinity(podSpec.Affinity, sfs)
 	currentSet.Spec.Template.Spec.Volumes = sfsVolume.Volumes
 
-	return r.client.Update(context.TODO(), currentSet)
+	err = r.client.Update(context.TODO(), currentSet)
+	if err != nil {
+		return fmt.Errorf("update error: %v", err)
+	}
+
+	if !cr.Spec.SmartUpdateEnabled() {
+		return nil
+	}
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: currentSet.Name, Namespace: currentSet.Namespace}, currentSet)
+	if err != nil {
+		return fmt.Errorf("failed to get sate: %v", err)
+	}
+
+	return r.smartUpdate(sfs, cr)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api.PerconaXtraDBCluster) error {
+	if !isPXC(sfs) {
+		return nil
+	}
+
+	if sfs.StatefulSet().Status.UpdatedReplicas >= sfs.StatefulSet().Status.Replicas {
+		return nil
+	}
+
+	if sfs.StatefulSet().Status.ReadyReplicas < sfs.StatefulSet().Status.Replicas {
+		return fmt.Errorf("failed to run 'SmartUpdate': not all replicas are ready")
+	}
+
+	list := corev1.PodList{}
+	err := r.client.List(context.TODO(),
+		&client.ListOptions{
+			Namespace:     sfs.StatefulSet().Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+		&list,
+	)
+	if err != nil {
+		return fmt.Errorf("get pod list: %v", err)
+	}
+
+	log.Info("statefullSet was changed, run smart update")
+
+	primary, err := r.getPrimaryPod(cr)
+	if err != nil {
+		return fmt.Errorf("get primary pod: %v", err)
+	}
+
+	log.Info(fmt.Sprintf("primary pod is %s", primary))
+
+	waitLimit := 120
+	if cr.Spec.PXC.LivenessInitialDelaySeconds != nil {
+		waitLimit = int(*cr.Spec.PXC.LivenessInitialDelaySeconds)
+	}
+
+	sort.Slice(list.Items, func(i, j int) bool {
+		return list.Items[i].Name > list.Items[j].Name
+	})
+
+	var primaryPod *corev1.Pod
+	for _, pod := range list.Items {
+		pod := pod
+		if strings.HasPrefix(primary, fmt.Sprintf("%s.%s.%s", pod.Name, sfs.StatefulSet().Name, sfs.StatefulSet().Namespace)) {
+			primaryPod = &pod
+		} else {
+			log.Info(fmt.Sprintf("delete secondary pod %s", pod.Name))
+			err := r.client.Delete(context.TODO(), &pod)
```

Comment:
image situation when we have a database which so big that it cannot start in 120 seconds.
in such case our SmartUpdate will constantly delete the first non-primary Pod on each reconciles loop (when pod finally started).

Idea: create `checkAndDelete(sts.updateRevision, &pod) ` function that check Pod's `controller-revision-hash` and remove Pod if `pod.meta.labels.controller-revision-hash != sts.updateRevision`

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/369/commits/40d38a6a749f75f9ec34fd2aa25fa9891b578c88